File: tests/fixtures/java_repo/src/main/java/com/example/api/services/ItemService.java
package com.example.api.services;

import com.example.api.dto.CreateItemRequest;
import com.example.api.dto.ItemResponse;
import com.example.api.dto.UpdateItemRequest;
import com.example.api.exceptions.ApiException;
import com.example.api.models.Item;
import com.example.api.repositories.ItemRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;

/**
 * ItemService — CRUD operations for {@link Item} aggregates.
 *
 * <p>Every read and write enforces ownership: items only come back to
 * the user who owns them, and {@code update}/{@code delete} on someone
 * else's item returns 403 Forbidden.
 */
@Service
public class ItemService {

    private static final int MAX_PAGE_SIZE = 100;

    private final ItemRepository itemRepository;

    public ItemService(ItemRepository itemRepository) {
        this.itemRepository = itemRepository;
    }

    /**
     * createItem persists a new item owned by the supplied user.
     */
    @Transactional
    public ItemResponse createItem(Long ownerId, CreateItemRequest request) {
        Item item = new Item(ownerId, request.getTitle(), request.getDescription());
        Item saved = itemRepository.save(item);
        return ItemResponse.fromEntity(saved);
    }

    /**
     * getItemById fetches a single item and enforces ownership.
     */
    public ItemResponse getItemById(Long ownerId, Long itemId) {
        Item item = loadOwned(ownerId, itemId);
        return ItemResponse.fromEntity(item);
    }

    /**
     * listItems returns a page of items owned by the given user.
     */
    public List<ItemResponse> listItems(Long ownerId, int page, int pageSize) {
        int safePage = Math.max(0, page);
        int safeSize = Math.min(MAX_PAGE_SIZE, Math.max(1, pageSize));
        Pageable pageable = PageRequest.of(safePage, safeSize);
        Page<Item> rows = itemRepository.findByOwnerId(ownerId, pageable);
        return rows.stream().map(ItemResponse::fromEntity).toList();
    }

    /**
     * updateItem applies the non-null fields of {@code patch} to an owned item.
     */
    @Transactional
    public ItemResponse updateItem(Long ownerId, Long itemId, UpdateItemRequest patch) {
        Item item = loadOwned(ownerId, itemId);
        if (patch.getTitle() != null) {
            item.setTitle(patch.getTitle());
        }
        if (patch.getDescription() != null) {
            item.setDescription(patch.getDescription());
        }
        return ItemResponse.fromEntity(item);
    }

    /**
     * deleteItem removes an owned item or 404/403s.
     */
    @Transactional
    public void deleteItem(Long ownerId, Long itemId) {
        int affected = itemRepository.deleteByIdAndOwnerId(itemId, ownerId);
        if (affected == 0) {
            throw new ApiException(HttpStatus.NOT_FOUND, "item not found");
        }
    }

    /**
     * loadOwned fetches an item and verifies the supplied user owns it.
     * Throws 404 when the item is missing, 403 when ownership mismatches.
     */
    private Item loadOwned(Long ownerId, Long itemId) {
        Item item = itemRepository.findById(itemId)
            .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "item not found"));
        if (!item.isOwnedBy(ownerId)) {
            throw new ApiException(HttpStatus.FORBIDDEN, "forbidden");
        }
        return item;
    }
}

File: tests/fixtures/cpp_repo/src/services/item_service.cpp
#include "api/services/item_service.hpp"

#include "api/error/api_error.hpp"
#include "api/repository/item_repository.hpp"

#include <algorithm>
#include <chrono>
#include <utility>

namespace api {

namespace {

std::int64_t now_seconds() {
    using namespace std::chrono;
    return duration_cast<seconds>(system_clock::now().time_since_epoch()).count();
}

}  // namespace

ItemService::ItemService(std::shared_ptr<ItemRepository> items)
    : items_(std::move(items)) {}

ItemResponse ItemService::create_item(std::int64_t owner_id,
                                      const CreateItemRequest& request) {
    if (request.title.empty()) {
        throw ApiError::bad_request("title must not be empty");
    }
    Item item{};
    item.owner_id = owner_id;
    item.title = request.title;
    item.description = request.description;
    item.created_at = now_seconds();
    Item saved = items_->save(item);
    return ItemResponse::from_entity(saved);
}

ItemResponse ItemService::get_item(std::int64_t requesting_user,
                                   std::int64_t item_id) {
    auto item = items_->find_by_id(item_id);
    if (!item.has_value()) {
        throw ApiError::not_found("item not found");
    }
    if (!item->is_owned_by(requesting_user)) {
        throw ApiError::forbidden("item is owned by another user");
    }
    return ItemResponse::from_entity(*item);
}

std::vector<ItemResponse> ItemService::list_items_for_owner(
    std::int64_t owner_id, int page, int page_size) {
    int safe_page = std::max(0, page);
    int safe_size = std::clamp(page_size, 1, kMaxPageSize);
    auto rows = items_->find_by_owner(owner_id, safe_page, safe_size);
    std::vector<ItemResponse> out;
    out.reserve(rows.size());
    for (const auto& item : rows) {
        out.push_back(ItemResponse::from_entity(item));
    }
    return out;
}

ItemResponse ItemService::update_item(std::int64_t requesting_user,
                                      std::int64_t item_id,
                                      const UpdateItemRequest& patch) {
    auto item = items_->find_by_id(item_id);
    if (!item.has_value()) {
        throw ApiError::not_found("item not found");
    }
    if (!item->is_owned_by(requesting_user)) {
        throw ApiError::forbidden("item is owned by another user");
    }
    if (patch.title.has_value()) {
        item->title = *patch.title;
    }
    if (patch.description.has_value()) {
        item->description = *patch.description;
    }
    Item saved = items_->update(*item);
    return ItemResponse::from_entity(saved);
}

void ItemService::delete_item(std::int64_t requesting_user, std::int64_t item_id) {
    auto item = items_->find_by_id(item_id);
    if (!item.has_value()) {
        throw ApiError::not_found("item not found");
    }
    if (!item->is_owned_by(requesting_user)) {
        throw ApiError::forbidden("item is owned by another user");
    }
    items_->delete_by_id(item_id);
}

}  // namespace api
